mopidy_tidal/library.py

```diff
@@ -157,104 +151,111 @@ def search(self, query=None, uris=None, exact=False):
-                            playlist = session.get_playlist(playlist_id)
-                            img_uri = playlist.image
-                            self.lru_playlist_img[playlist_id] = img_uri
-
-                        uri_images = [Image(uri=img_uri, width=512, height=512)]
-                    elif parts[1] == 'track':
-                        album_id = parts[3]
-                        img_uri = self.lru_album_img.hit(album_id)
-                        if img_uri is None:
-                            album = session.get_album(album_id)
-                            img_uri = album.image
-                            self.lru_album_img[album_id] = img_uri
-
-                        uri_images = [Image(uri=img_uri, width=512, height=512)]
-                        pass
-
-                images[uri] = uri_images or ()
-            except AttributeError:
-                logger.error("AttributeError when processing URI %r" % uri)
-            except HTTPError as err:
-                logger.error("HTTPError when processing URI %r" % uri)
+                images[uri] = self._get_image(uri)
+            except (AssertionError, AttributeError, HTTPError) as err:
+                logger.error(
+                    "%s when processing URI %r: %s",
+                    type(err), uri, err)
 
+        self._image_cache.update(images)
         return images
 
     def lookup(self, uris=None):
         logger.info("Lookup uris %r", uris)
-        session = self.backend._session
         if isinstance(uris, str):
             uris = [uris]
         if not hasattr(uris, '__iter__'):
             uris = [uris]
 
         tracks = []
-        for uri in uris:
+        for uri in (uris or []):
+            data = []
             try:
                 parts = uri.split(':')
-                if uri.startswith('tidal:track:'):
-                    if uri in self.track_cache:
-                        tracks.append(self.track_cache[uri])
-                    else:
-                        tracks += self._lookup_track(session, parts)
-                elif uri.startswith('tidal:album'):
-                    tracks += self._lookup_album(session, parts)
-                elif uri.startswith('tidal:artist'):
-                    tracks += self._lookup_artist(session, parts)
-                elif uri.startswith('tidal:playlist'):
-                    tracks += self._lookup_playlist(session, parts)
-
-            except AttributeError:
-                logger.error("AttributeError when processing URI %r" % uri)
-            except HTTPError as err:
-                logger.error("HTTPError when processing URI %r" % uri)
+                cache = getattr(self, f'_{parts[1]}_cache')
+                lookup = getattr(self, f'_lookup_{parts[1]}')
+
+                try:
+                    data = cache[uri]
+                except KeyError:
+                    data = lookup(self._session, parts)
+
+                tracks += data if hasattr(data, '__iter__') else [data]
```

Comment:
I think this line should be `tracks += data if hasattr(data, '__iter__') else data.tracks` to fix the error `Expected a list of Track, not [Playlist..` when refreshing playlists

Reply:
I have fixed the issue by letting `_lookup_playlist` return a `tracks, playlist` tuple instead of a list of tracks. For lookup purposes we need the list, but for caching purposes we need the playlist object.

Reply:
@tehkillerbee this should be fixed in https://github.com/tehkillerbee/mopidy-tidal/pull/60/commits/2adfb83435a9e7196fef85891ca77efb3d84c749